packages/ifpack2/src/Ifpack2_Details_Fildl_def.hpp

```diff
@@ -0,0 +1,282 @@
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+//@HEADER
+*/
+
+/// @file Ifpack2_fildl_def.hpp
+
+#ifndef __IFPACK2_FLDL_DEF_HPP__ 
+#define __IFPACK2_FLDL_DEF_HPP__ 
+
+#include <Ifpack2_Details_Fildl_decl.hpp>
+#include <Ifpack2_Details_CrsArrays.hpp>
+
+namespace Ifpack2
+{
+namespace Details
+{
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+Fildl<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+Fildl(Teuchos::RCP<const TRowMatrix> mat_)
+{
+  mat = mat_;
+  initFlag = false;
+  computedFlag = false;
+  nApply = 0;
+  nComputed = 0;
+  nInit = 0;
+  initTime = 0.0;
+  computeTime = 0.0;
+  applyTime = 0.0;
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+Teuchos::RCP<const Tpetra::Map<LocalOrdinal,GlobalOrdinal,Node> >
+Fildl<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getDomainMap () const
+{
+  return mat->getDomainMap();
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+Teuchos::RCP<const Tpetra::Map<LocalOrdinal,GlobalOrdinal,Node> >
+Fildl<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getRangeMap () const
+{
+  return mat->getRangeMap();
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void Fildl<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+apply (const Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node> &X,
+    Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node> &Y,
+    Teuchos::ETransp mode,
+    Scalar alpha,
+    Scalar beta) const
+{
+  if(!initFlag || !computedFlag)
+  {
+    throw std::runtime_error("Called Fic::apply() without first calling initialize() and/or compute().");
+  }
+  if(X.getNumVectors() != Y.getNumVectors())
+  {
+    throw std::invalid_argument("Filu::apply: X and Y have different numbers of vectors.");
+  }
+  int nvecs = X.getNumVectors();
+  ScalarArray x;
+  ScalarArray y;
+  auto x2D = X.template getLocalView<ExecSpace>();
+  auto y2D = Y.template getLocalView<ExecSpace>();
+  //Solve each vector one at a time (until FastILU supports multiple RHS)
+  for(int i = 0; i < nvecs; i++)
+  { 
+    auto x1D = Kokkos::subview(x2D, Kokkos::ALL(), i);
```

Comment:
You may, but please see #1470.  I would do this: If the vector has only one column, the above code is fine.  Otherwise, for now, you could use `getVector(j)`, but see #1470.  Note that getLocalView() always returns a 2-D View, so you'll still have to do `Kokkos::subview(X_j, Kokkos::ALL (), 0)` on it.